### ISSUE TITLE
IP lookup failed: couldn't find MAC address in VMX

### ISSUE THREAD

**Issue by nbayle** [hashicorp/packer#5925]
I'm trying to use vagrant ubuntu 16.04 bento box as a source image with the vmware-vmx builder, but it continually cannot find the MAC address and, subsequently, the IP address.

First off, is using a vagrant box as a source generally frowned upon?

Full packer log:
https://gist.github.com/nbayle/d9fc70f492d01c6a5acceb2454b8fa67

Snippet:

```
2018/02/20 11:52:13 packer: 2018/02/20 11:52:13 Executing: /usr/bin/vmrun -T ws start output-vmware-vmx/packer-vmware-vmx-1519156314.vmx nogui
2018/02/20 11:52:14 packer: 2018/02/20 11:52:14 stdout:
2018/02/20 11:52:14 packer: 2018/02/20 11:52:14 stderr:
2018/02/20 11:52:14 ui: ==> vmware-vmx: Waiting 10s for boot...
==> vmware-vmx: Waiting 10s for boot...
==> vmware-vmx: Connecting to VM via VNC (127.0.0.1:5923)
2018/02/20 11:52:24 ui: ==> vmware-vmx: Connecting to VM via VNC (127.0.0.1:5923)
2018/02/20 11:52:24 packer: 2018/02/20 11:52:24 Connected to VNC desktop: Clone of ubuntu-16.04-amd64
2018/02/20 11:52:24 packer: 2018/02/20 11:52:24 Host IP for the VMware machine: 192.168.2.1
==> vmware-vmx: Typing the boot command over VNC...
2018/02/20 11:52:24 ui: ==> vmware-vmx: Typing the boot command over VNC...
2018/02/20 11:52:24 packer: 2018/02/20 11:52:24 [INFO] Waiting for SSH, up to timeout: 10m0s
2018/02/20 11:52:24 ui: ==> vmware-vmx: Waiting for SSH to become available...
==> vmware-vmx: Waiting for SSH to become available...
2018/02/20 11:52:24 packer: 2018/02/20 11:52:24 Lookup up IP information...
2018/02/20 11:52:24 packer: 2018/02/20 11:52:24 IP lookup failed: couldn't find MAC address in VMX
2018/02/20 11:52:24 packer: 2018/02/20 11:52:24 [DEBUG] Error getting SSH address: IP lookup failed: couldn't find MAC address in VMX
2018/02/20 11:52:29 packer: 2018/02/20 11:52:29 Lookup up IP information...
2018/02/20 11:52:29 packer: 2018/02/20 11:52:29 IP lookup failed: couldn't find MAC address in VMX
2018/02/20 11:52:29 packer: 2018/02/20 11:52:29 [DEBUG] Error getting SSH address: IP lookup failed: couldn't find MAC address in VMX
```

Misc:
```
{
  "variables": {
    "home": "{{env `HOME`}}"
  },
  "builders": [{
    "type": "vmware-vmx",
    "source_path": "{{user `home`}}/.vagrant.d/boxes/bento-VAGRANTSLASH-ubuntu-16.04/201802.02.0/vmware_desktop/ubuntu-16.04-amd64.vmx",
    "headless": "true",
    "boot_wait": "10s",
    "ssh_username": "vagrant",
    "ssh_password": "vagrant",
    "ssh_wait_timeout": "10m",
    "shutdown_command": "echo 'vagrant' | sudo -S shutdown -h 0",
    "vmx_data": {
      "memsize": "2048"
    }
  }],
  "provisioners": [
    {
      "type": "chef-solo",
      "cookbook_paths": ["cookbooks"],
      "run_list": ["akm"]
    },
    {
      "type": "shell",
      "inline": ["echo 'vagrant' | sudo apt-get --yes --purge autoremove chef"]
    }
  ]
}

packer --version
1.2.1

cat /etc/debian_version 
8.10
```

**Comment by arizvisa** [comment]
Hi @nbayle, do you know if it works with an older version of packer like v1.1.3? (i.e. a regression)

To build an older version you can follow [CONTRIBUTING.md#setting-up-go-to-work-on-packer](../../packer/blob/master/CONTRIBUTING.md#setting-up-go-to-work-on-packer) or do something like:
```
$ go get github.com/hashicorp/packer
$ cd ~/.go/src/github.com/hashicorp/packer
$ git checkout v1.1.3
$ make deps
$ make dev
```

**Comment by nbayle** [comment]
1.1.3 gives the same error. This is my first time using the vmware-vmx builder, so I've never had it work before.

Full log:
https://gist.github.com/nbayle/d9fa2687c001f5309a175cfc84d58c05

**Comment by rickard-von-essen** [comment]
Double check that there is actually a NIC in the VM.

**Comment by nbayle** [comment]
So, there's no NIC (ethernet0) in the VMX file. However, the VM fires up fine in vagrant using vmware, which I imagine creates the NIC on startup. So is this just a difference in how the apps utilize vmware workstation or do I have to generate the NIC manually (if so, is there a vmware command that can just generate an interface)?

**Comment by arizvisa** [comment]
Hmm, maybe adding a `vmx_add_ethernet_interfaces` to  the vmware-vmx builder to counter the `vmx_remove_ethernet_interfaces` option would be useful in this situation?

Not sure what'd be a good way to predict the hwaddr to put for `ethernet0`, though. Although, now that there's a way to parse the hwaddrs in use by local vmware instances, we could probably generate it and then check it against what's in the used dhcp leases... It just might require some fuckery if the esx5 driver in packer is used since I'm not sure how to enumerate hwaddrs there...

Any opinions?

**Comment by rickard-von-essen** [comment]
I think this just works if you add a NIC to the VMX. Packer doesn't modify your VM but expect that there is a NIC. Vagrant on the other hand have many expectations on a box and alot of manipulations of your VM depending on your Vagrantfile. 

`vmware-vmx` doesn't support remote (ESXi).

**Comment by nbayle** [comment]
Adding:

```
"ethernet0.present": "true",
"ethernet0.startConnected": "true",
"ethernet0.vnet": "vmnet8"
```

To `vmx_data` in my packer config works to get rid of the "couldn't find MAC address in VMX" error.

The new error is:
```
Error getting SSH address: IP lookup failed: IP not found for MAC 00:0c:29:e7:ce:70 in DHCP leases at /etc/vmware/vmnet8/dhcpd/dhcpd.leases
```

And indeed, the dhcpd.leases file is empty. That said, the VM _does_ get an IP address from the dhcp server in vmware:
![screen shot 2018-02-22 at 13 10 46](https://user-images.githubusercontent.com/8433050/36564570-9de7a286-17d2-11e8-85d7-d5184b1a6442.png)

Anyone seen this before?

Thank you.

**Comment by mwhooker** [comment]
@nbayle are you able to test the latest version of packer from master?

**Comment by nbayle** [comment]
@mwhooker This was using 1.2.1 that someone built the other day. But If you need me to pull fresh, I think I've got building it figured out.

**Comment by mwhooker** [comment]
hmm, I was wondering if it was fixed by #5900, but not sure it's related.

Are you still getting failures on 1.1.3 now that you've changed `vmx_data`?

**Comment by nbayle** [comment]
Same error with 1.1.3

```2018/02/22 15:42:40 packer: 2018/02/22 15:42:40 IP lookup failed: IP not found for MAC 00:0c:29:e9:a6:7b in DHCP leases at /etc/vmware/vmnet8/dhcpd/dhcpd.leases```

**Comment by nbayle** [comment]
Digging deeper, you can see that it _should_ be writing to the leases file if you look at the `ps` output:

`11899 ?        Ss     0:00 /usr/bin/vmnet-dhcpd -s 6 -cf /etc/vmware/vmnet8/dhcpd/dhcpd.conf -lf /etc/vmware/vmnet8/dhcpd/dhcpd.leases -pf /var/run/vmnet-dhcpd-vmnet8.pid vmnet8`

**Comment by nbayle** [comment]
I opened a support ticket with VMware to see if they have any idea why leases aren't being written.

**Comment by nbayle** [comment]
VMware's support, in addition to being useless, refuses to acknowledge that dhcp leases are written to the dhcpd.leases file at all and states this is normal operation.

Does anyone have a contact at VMware that can look into this to see if it's a bug?

**Comment by nbayle** [comment]
The plot thickens. Apparently I'm getting an IP address from an upstream DHCP server outside VMware. So, I went to switch to hostonly/vmnet1 and it appears that directive is being ignored?

```
    "vmx_data": {
      "memsize": "2048",
      "ethernet0.connectionType": "hostonly",
      "ethernet0.vnet":  "vmnet1"
    }
```

Yet, the output from packer shows the following:
```
2018/03/20 11:30:54 packer: 2018/03/20 11:30:54 Executing: /usr/bin/vmrun -T ws clone /home/nicholas/.vagrant.d/boxes/bento-VAGRANTSLASH-ubuntu-16.04/201802.02.0/vmware_desktop/ubuntu-16.04-amd64.vmx output-vmware-vmx/packer-vmware-vmx-1521570654.vmx full
2018/03/20 11:30:57 packer: 2018/03/20 11:30:57 stdout:
2018/03/20 11:30:57 packer: 2018/03/20 11:30:57 stderr:
2018/03/20 11:30:57 packer: 2018/03/20 11:30:57 Defaulting to network type : nat
2018/03/20 11:30:57 packer: 2018/03/20 11:30:57 Setting VMX: 'memsize' = '2048'
2018/03/20 11:30:57 packer: 2018/03/20 11:30:57 Setting VMX: 'ethernet0.connectionType' = 'hostonly'
2018/03/20 11:30:57 packer: 2018/03/20 11:30:57 Setting VMX: 'ethernet0.vnet' = 'vmnet1'
2018/03/20 11:30:57 packer: 2018/03/20 11:30:57 Writing VMX to: output-vmware-vmx/packer-vmware-vmx-1521570654.vmx
[ snip ]
==> vmware-vmx: Waiting for SSH to become available...
2018/03/20 11:31:28 ui: ==> vmware-vmx: Waiting for SSH to become available...
2018/03/20 11:31:28 packer: 2018/03/20 11:31:28 MEGAN: NEtworkMapper is common.NetworkMap{map[string]string{"name":"Bridged", "device":"vmnet0"}, map[string]string{"name":"HostOnly", "device":"vmnet1"}, map[string]string{"name":"NAT", "device":"vmnet8"}}
2018/03/20 11:31:28 packer: 2018/03/20 11:31:28 MEGAN: network is "nat"
2018/03/20 11:31:28 packer: 2018/03/20 11:31:28 Lookup up IP information...
2018/03/20 11:31:28 packer: 2018/03/20 11:31:28 IP lookup failed: couldn't find MAC address in VMX
```

I'm assuming that shouldn't read "network is 'nat'".

**Comment by SwampDragons** [comment]
Speaking to your original question which somehow I missed earlier:  "is using a vagrant box as a source generally frowned upon" not really; a vagrant box is just a compressed folder with the vmx and some config information. If you're linking directly to the vmx and the box is uncompressed, then there should not be any issue. Though your vagrant parts of it will be ignored.

**Comment by SwampDragons** [comment]
What happens if you bump the ssh_wait_timeout to 20m?  I've seen it take a long time for the IP to appear in the leases.dhcpd file

**Comment by SwampDragons** [comment]
Skimming through the code, I was unable to immediately see where the vmx builder was actually using the user-input vmx data; that doesn't mean it isn't there, just that I'm not super familiar with this code and it's buried under some layers of abstraction. But it certainly doesn't seem to be used based on that logline. I guess an easy way of finding out is by directly modifying the vmx file you're passing into packer, rather than adding modifications in the Packer config.

**Comment by SwampDragons** [comment]
@arizvisa any ideas here? I suspect at this point you're more familiar with the vmware workstation code than I.

**Comment by arizvisa** [comment]
Sorry, I don't have too much time to look in detail at this until this weekend for sure. Is that okay with your guys schedule?

The custom VMX data is added to the .vmx by `StepConfigureVMX` in `builder/vmware/common/step_configure_vmx.go:52`. It's essentially adding any user-specified keys right into the .vmx file.

According to `builder/vmware/iso/builder.go`, that `StepConfigureVMX` code is happening right _after_ `StepCreateVMX` (where the network parsing code is happening). So, it seems like the logic would be that anything overwritten by `StepConfigureVMX` should take priority over `StepCreateVMX` (excluding case sensitivity maybe?).

If for some reason it's not overwriting it, there's a chance that VMware might be giving priority to another instance of `ethernet0.connectionType`. If that's the actual case, then I can probably explicitly check to see if a user manually specified the connectionType and to honor that instead or to figure out why the user-input vmx data is not getting priority.

@nbayle: Can you supply your latest .vmx that packer generates? Does it include your custom vmx_data?

**Comment by arizvisa** [comment]
Oh interesting, it also looks like StepConfigureVMX is actually running twice. Once for the VMXData and another for the VMXDataPost. One of them should overwrite it though...

But yea, if the theory I suggested is the issue we can add checks to `stepCreateVMX` to read the explicit VMXData when determining the default network instead of depending on it being overwritten by `StepConfigureVMX`.

**Comment by nbayle** [comment]
Yeah, the weekend is fine by me.

So I edited the vmx file in the box and added the ethernet0 statements. Then I fired up packer and looked at the vmx file it created...

packer output:
```
2018/03/21 15:44:43 packer: 2018/03/21 15:44:43 MEGAN: NEtworkMapper is common.NetworkMap{map[string]string{"device":"vmnet0", "name":"Bridged"}, map[string]string{"name":"HostOnly", "device":"vmnet1"}, map[string]string{"device":"vmnet8", "name":"NAT"}}
2018/03/21 15:44:43 packer: 2018/03/21 15:44:43 MEGAN: network is "nat"
```

The output-vmware-vmx/packer-vmware-vmx-1521672214.vmx:
```
ethernet0.connectiontype = "hostonly"
ethernet0.vnet = "vmnet1"
```

Here's the whole vmx that packer generates, which looks good to me as far as ethernet0:
https://gist.github.com/nbayle/da0a4e472d20aeea687d280746196dc5

**Comment by nbayle** [comment]
Are vmx files case-sensitive? After the vmx gets run through packer, `connectionType` becomes `connectiontype`.

**Comment by SwampDragons** [comment]
I know [this](https://github.com/hashicorp/packer/pull/5985) was supposedly for vmware fusion, but I'm wondering if it's related somehow.

**Comment by nbayle** [comment]
I'll give it another go when 1.2.2 rolls out.

**Comment by SwampDragons** [comment]
That sounds fair. Enough has changed recently with related-enough seeming code that this may already be fixed. We should be releasing 1.2.2 tomorrow, so you don't have long to wait. Sorry this has been such a pain.

**Comment by nbayle** [comment]
That's fantastic. Forward momentum is all I care about....

**Comment by arizvisa** [comment]
@nbayle, wrt to case sensitivity it doesn't appear to be case-sensitive on VMware Workstation.

I added `ethernet0.present = "TRUE"` to the .vmx and Workstation seems to not care about the case sensitivity.

**Comment by arizvisa** [comment]
So the reason why the logging outputs "network is nat" is actually because of a case-sensitivity issue with ParseVMX which results in the following lookup not finding the network type. This is only in the vmware-vmx builder though. I'll have a PR in a bit that'll look for all cases of vmxData being accessed and then adjusting the case-sensitivity so that it's lowercase.

Although, the custom vmx data should still be taking priority, so this won't fix that. I'll still dig though..

```
	var networkType string
	if _, ok := vmxData["ethernet0.connectionType"]; ok {
		networkType = vmxData["ethernet0.connectionType"]
	}
	if networkType == "" {
		networkType = "nat"
		log.Printf("Defaulting to network type : nat")
	}
```

**Comment by arizvisa** [comment]
@SwampDragons, created a PR #6057 that fixes that case sensitivity issue i mentioned when determining the network type for the vmware-vmx builder.

PR #5985 is unfortunately only for the vmware-iso builder, so I don't think it influences this issue.

**Comment by arizvisa** [comment]
Despite the network being determined incorrectly due to the case-sensitivity issue, the custom `vmx_data` is still being honored and is overwriting `ethernet0.connectionType` in the provided .vmx.

In my tests with Workstation (I don't have Fusion), it's choosing the correct network type based on whether it's "hostonly" or "nat". So Fusion acting as though "bridged" is being specified is pretty confusing unless the "vmnet8" interface itself is "bridged".

**Comment by nbayle** [comment]
Hmm. I'm not using Fusion either. Are you using workstation version 14?

**Comment by arizvisa** [comment]
@nbayle, yep. 14.0.0 build-6661328

**Comment by SwampDragons** [comment]
@nbayle the 1.2.2 binary is released now... could you try with that and let me know if the other vmware-networking related changes resolved this issue too?

**Comment by nbayle** [comment]
Apologies, tried it last night and no change. Pretty confused though if arizvisa has it working...

**Comment by arizvisa** [comment]
@nbayle, so just to confirm: That .vmx you gave me being used outside of packer appears as a bridged interface in VMware?

**Comment by nbayle** [comment]
I've not tried it outside of packer, but if I login with VNC, while it's "waiting for ssh" and run ifconfig, the IP address that it has is the same as if it were bridged (192.168.99.xxx) and shows up on the dhcp server log outside of vmware. Whereas the nat vmnet8 dhcp server should issue a 192.168.2.xxx and the hostonly vmnet1 dhcp server should issue a 172.16.227.xxx address.

This behavior doesn't change no matter what I put in the vmx itself or in the packer config.

**Comment by nbayle** [comment]
I'm currently creating a fresh system to start over, since you had success. Maybe something about that environment is wrong.

**Comment by nbayle** [comment]
OK, I rebuilt my whole environment and it's working using NAT right now.

That said, it will still ignore any vmnet1 or hostnonly directives that I add in my packer json config (which I don't need). It'd be nice to get that fixed for the sake of fixing, but other than that, I'm good to go for now.

Thank you all for your work on this.
Nick

**Comment by arizvisa** [comment]
Does it work via the new "network" configuration for vmware-iso?

**Comment by arizvisa** [comment]
Sorry, let me rephrase that: If you try using "network" set to hostonly, does it work?

**Comment by nbayle** [comment]
I'm using vmware-vmx. Does vmx have the new network config?

**Comment by arizvisa** [comment]
Ah no it doesn't. :-/ vmware-iso only. So nevermind I guess.

**Comment by SwampDragons** [comment]
Closing since it appears the original issue was solved some time ago.

**Comment by alleynec** [comment]
I'm actually getting the exact same error using Packer 1.3.1 and after trying all above suggestions. Below is my vmx file (I don't use the vmx_data or vmx_data_post fields in the vmx builder section of the json file):

.encoding = "UTF-8"
cleanShutdown = "TRUE"
config.version = "8"
cpuid.coresPerSocket = "4"
displayName = "Win-8.1-x64-base"
ehci.pciSlotNumber = "34"
ehci.present = "TRUE"
ethernet0.connectionType = "nat"
ethernet0.present = "TRUE"
ethernet0.startConnected = "TRUE"
extendedConfigFile = "Win-8.1-x64-base.vmxf"
floppy0.present = "FALSE"
guestOS = "windows8-64"
hpet0.present = "TRUE"
keyboardAndMouseProfile = "520c140c-fca4-0ebb-4d54-f6ca38630681"
mem.hotadd = "TRUE"
memsize = "3072"
migrate.hostlog = "./Win-8.1-x64-base-afb40be1.hlog"
mks.enable3d = "TRUE"
monitor.phys_bits_used = "43"
numa.autosize.cookie = "40001"
numa.autosize.vcpu.maxPerVirtualNode = "4"
numvcpus = "4"
nvram = "Win-8.1-x64-base.nvram"
pciBridge0.pciSlotNumber = "17"
pciBridge0.present = "TRUE"
pciBridge4.functions = "8"
pciBridge4.pciSlotNumber = "21"
pciBridge4.present = "TRUE"
pciBridge4.virtualDev = "pcieRootPort"
pciBridge5.functions = "8"
pciBridge5.pciSlotNumber = "22"
pciBridge5.present = "TRUE"
pciBridge5.virtualDev = "pcieRootPort"
pciBridge6.functions = "8"
pciBridge6.pciSlotNumber = "23"
pciBridge6.present = "TRUE"
pciBridge6.virtualDev = "pcieRootPort"
pciBridge7.functions = "8"
pciBridge7.pciSlotNumber = "24"
pciBridge7.present = "TRUE"
pciBridge7.virtualDev = "pcieRootPort"
powerType.powerOff = "soft"
powerType.powerOn = "soft"
powerType.reset = "soft"
powerType.suspend = "soft"
proxyApps.publishToHost = "FALSE"
sata0:0.fileName = "Virtual Disk-000018.vmdk"
sata0:0.present = "TRUE"
sata0:0.redo = ""
sata0:1.deviceType = "cdrom-image"
sata0:1.fileName = "/VMware/Installers/en_windows_8.1_enterprise_with_update_x64_dvd_6054382.iso"
sata0:1.present = "TRUE"
sata0:1.startConnected = "FALSE"
sata0.pciSlotNumber = "36"
sata0.present = "FALSE"
scsi0:0.present = "FALSE"
scsi0.pciSlotNumber = "160"
scsi0.present = "TRUE"
scsi0.sasWWID = "50 05 05 68 d4 41 3e f0"
scsi0.virtualDev = "lsisas1068"
sensor.location = "pass-through"
serial0.fileType = "thinprint"
serial0.present = "TRUE"
serial0.startConnected = "FALSE"
softPowerOff = "TRUE"
sound.autodetect = "TRUE"
sound.fileName = "-1"
sound.pciSlotNumber = "33"
sound.present = "TRUE"
sound.virtualDev = "hdaudio"
svga.graphicsMemoryKB = "786432"
tools.syncTime = "TRUE"
tools.upgrade.policy = "upgradeAtPowerCycle"
usb_xhci:4.deviceType = "hid"
usb_xhci:4.parent = "-1"
usb_xhci:4.port = "4"
usb_xhci:4.present = "TRUE"
usb_xhci:6.deviceType = "hub"
usb_xhci:6.parent = "-1"
usb_xhci:6.port = "6"
usb_xhci:6.present = "TRUE"
usb_xhci:6.speed = "2"
usb_xhci:7.deviceType = "hub"
usb_xhci:7.parent = "-1"
usb_xhci:7.port = "7"
usb_xhci:7.present = "TRUE"
usb_xhci:7.speed = "4"
usb_xhci.pciSlotNumber = "224"
usb_xhci.present = "TRUE"
usb.pciSlotNumber = "32"
usb.present = "TRUE"
usb.vbluetooth.startConnected = "FALSE"
uuid.bios = "56 4d 12 b8 d4 41 3e fe-cf 9c a3 97 e0 18 58 ce"
uuid.location = "56 4d 73 98 1a 8b 33 6d-52 e7 84 d0 28 b5 85 c0"
virtualHW.productCompatibility = "hosted"
virtualHW.version = "14"
vm.genid = "3256754623135940363"
vm.genidX = "7760565860003738565"
vmci0.id = "-535275314"
vmci0.pciSlotNumber = "35"
vmci0.present = "TRUE"
vmotion.checkpointFBSize = "134217728"
vmotion.checkpointSVGAPrimarySize = "134217728"
vmxnet3.serialNumberV2 = "FALSE"

**Comment by bcube** [comment]
I am receiving the same message. I am with packer 1.3.5. 
[pnda_template_base.vmx.txt](https://github.com/hashicorp/packer/files/3067999/pnda_template_base.vmx.txt)



2019/04/11 17:43:01 packer: 2019/04/11 17:43:01 Lookup up IP information...
2019/04/11 17:43:01 packer: 2019/04/11 17:43:01 GuestAddress found MAC address in VMX: 00:0c:29:64:ff:e1
2019/04/11 17:43:01 packer: 2019/04/11 17:43:01 Trying DHCP leases path: /etc/vmware/vmnet8/dhcpd/dhcpd.leases
2019/04/11 17:43:01 packer: 2019/04/11 17:43:01 IP lookup failed: None of the found device(s) [vmnet8] has a DHCP lease for MAC 00:0c:29:64:ff:e1
2019/04/11 17:43:01 packer: 2019/04/11 17:43:01 [DEBUG] Error getting SSH address: IP lookup failed: None of the found device(s) [vmnet8] has a DHCP lease for MAC 00:0c:29:64:ff:e1
2019/04/11 17:43:06 packer: 2019/04/11 17:43:06 Located networkmapper configuration file using Workstation: /etc/vmware/netmap.conf
2019/04/11 17:43:06 packer: 2019/04/11 17:43:06 GuestIP discovered device matching nat: vmnet8
